### PR TITLE
HOOK-584 Add default server which drops requests to unknown hosts

### PR DESCRIPTION
**Pull request by maikschwan**
Drop requests to unknown hosts by:
- adding a new server (first in list, first is the default for no match)
- return 444, nginx drops the connection to the client
- do not log those requests -> protect us from too many logs
- in case https is used: use a selfsigned dummy cert to resolve TLS and then reject the request

Testing and results:
✅ _Drop requests to ip instead of host_
```
curl -vk http://192.168.99.100/
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to 192.168.99.100 (192.168.99.100) port 80 (#0)
> GET / HTTP/1.1
> Host: 192.168.99.100
> User-Agent: curl/7.54.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host 192.168.99.100 left intact
curl: (52) Empty reply from server
```

✅ _Drop http request to unknown host_
```
curl -vk http://local-svc-1.leanix.net/
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to local-svc-1.leanix.net (192.168.99.100) port 80 (#0)
> GET / HTTP/1.1
> Host: local-svc-1.leanix.net
> User-Agent: curl/7.54.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host local-svc-1.leanix.net left intact
curl: (52) Empty reply from server
```

✅ _Drop http**s** request to unknown host_
```
curl -vk https://local-svc-1.leanix.net/
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to local-svc-1.leanix.net (192.168.99.100) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=DE; ST=Some-State; O=Internet Widgits Pty Ltd
*  start date: Jun 26 14:15:42 2019 GMT
*  expire date: Jul 26 14:15:42 2019 GMT
*  issuer: C=DE; ST=Some-State; O=Internet Widgits Pty Ltd
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
> GET / HTTP/1.1
> Host: local-svc-1.leanix.net
> User-Agent: curl/7.54.0
> Accept: */*
>
* TLSv1.2 (IN), TLS alert, Client hello (1):
* Empty reply from server
* Connection #0 to host local-svc-1.leanix.net left intact
curl: (52) Empty reply from server
```

✅ _http to https redirect still works for known hosts_
```
curl -vk http://local-svc.leanix.net/
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to local-svc.leanix.net (192.168.99.100) port 80 (#0)
> GET / HTTP/1.1
> Host: local-svc.leanix.net
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 302 Moved Temporarily
< Date: Wed, 26 Jun 2019 14:41:29 GMT
< Content-Type: text/html
< Content-Length: 154
< Connection: keep-alive
< Location: https://local-svc.leanix.net/
<
<html>
<head><title>302 Found</title></head>
<body bgcolor="white">
<center><h1>302 Found</h1></center>
<hr><center>nginx</center>
</body>
</html>
* Connection #0 to host local-svc.leanix.net left intact
```

✅ _https still works for known hosts (with leanix cert)_
```
curl -vk https://local-svc.leanix.net/
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to local-svc.leanix.net (192.168.99.100) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=DE; ST=Nordrhein-Westfalen; L=Bonn; O=LeanIX GmbH; CN=*.leanix.net
*  start date: Jul  3 00:00:00 2018 GMT
*  expire date: Aug 31 12:00:00 2020 GMT
*  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust TLS RSA CA G1
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: local-svc.leanix.net
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Wed, 26 Jun 2019 14:42:11 GMT
< Content-Type: text/html
< Content-Length: 449
< Last-Modified: Thu, 13 Jun 2019 09:12:47 GMT
< Connection: keep-alive
< ETag: "5d02138f-1c1"
< Accept-Ranges: bytes
<
<html>
  <head>
    <title>LeanIX</title>
    <style>
      body {
        background-color: #009fdf;
        color: #FFFFFF;
        font-family: Arial;
      }
      a, a:visited, a:focus {
        color: #FFFFFF;
      }
    </style>
  </head>
  <body>
    <h1>No content here</h1>
    Please enter the proper URL of your LeanIX Instance or ask for help at <a href="mailto:support@leanix.net">support@leanix.net</a>. Thank you.
  </body>
</html>
* Connection #0 to host local-svc.leanix.net left intact
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -18,6 +18,8 @@ COPY html /usr/share/nginx/html/
 COPY nginx.conf.tmpl /etc/confd/templates/nginx.conf.tmpl
 COPY nginx.toml /etc/confd/conf.d/nginx.toml
 COPY supervisord.conf /etc/supervisor/supervisord.conf
+COPY certs/selfsigned.pem /etc/nginx/selfsigned/cert.pem
```
Comment (romansaul) [inline Dockerfile]: When does this cert expire? How do we renew it? Why don't we use our wildcard cert for leanix.net?

File: nginx.conf.tmpl [hunk 2]
```diff
@@ -263,6 +263,17 @@ http {
     ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA:!3DES;
     ssl_prefer_server_ciphers on;
 
+    # Block requests to unkown hosts
+    server {
```
Comment (romansaul) [inline nginx.conf.tmpl]: Now I know why to omit server_name in this section! https://blog.gahooa.com/2013/08/21/nginx-how-to-specify-a-default-server/

File: Dockerfile (hunk 1 above)
Comment (romansaul) [reply Dockerfile]: Your test lists:
*  start date: Jun 26 14:15:42 2019 GMT
*  expire date: Jul 26 14:15:42 2019 GMT

What does this mean?

Comment (maikschwan) [reply Dockerfile]: The problem is how to ensure that the leanix wildcard cert is mounted so the nginx does not fail. For the known hosts this mechanism is done via https://github.com/leanix/docker-nginx-proxy/blob/feature/hook-584-disable-ip-and-unkown-host-access/nginx.conf.tmpl#L362
To ensure that the certificate is always in the right place, I think we would need to mount it separately.
Using this self-signed cert here allows us to leave the cert in the repo and copy it into the image, making this section more robust.
In terms of expiry, we could create a certificate with an expiry long in the future.

Comment (maikschwan) [reply Dockerfile]: I replaced the self signed certificate with a cert which expires in 30 years. This is the curl response:

```
curl -vk https://local-svc-1.leanix.net/
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to local-svc-1.leanix.net (192.168.99.100) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=leanix.net
*  start date: Jun 28 08:45:18 2019 GMT
*  expire date: Jun 20 08:45:18 2049 GMT
*  issuer: CN=leanix.net
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
> GET / HTTP/1.1
> Host: local-svc-1.leanix.net
> User-Agent: curl/7.54.0
> Accept: */*
>
* TLSv1.2 (IN), TLS alert, Client hello (1):
* Empty reply from server
* Connection #0 to host local-svc-1.leanix.net left intact
curl: (52) Empty reply from server
```

**Pull request merged**